src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,432 @@
+            while (!IsStatementOrFieldDeclaration(currentNode, syntaxFacts))
+            {
+                if (RequiredDelimiterIsMissing(currentNode))
+                {
+                    // A required delimiter is missing; do not treat semicolon as statement completion
+                    // Example: missing final `)` in `obj.Method($`
+                    return;
+                }
+
+                if (currentNode.Parent == null)
+                {
+                    return;
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            // if the statement syntax itself requires a closing delimiter, verify it is there
+            if (StatementClosingDelimiterIsMissing(currentNode))
+            {
+                // Example: missing final `)` in `do { } while (x$$`
+                return;
+            }
+
+            var semicolonPosition = GetSemicolonLocation(root, currentNode, caretPosition);
+
+            // Place cursor after the statement
+            args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(semicolonPosition));
+        }
+
+        private int GetSemicolonLocation(SyntaxNode root, SyntaxNode currentNode, int caretPosition)
+        {
+            int end;
+            if (currentNode.IsKind(SyntaxKind.ForStatement))
+            {
+                // in for statements, semicolon can go after initializer or after condition, depending on where the caret is located
+                var forStatementSyntax = (ForStatementSyntax)currentNode;
+                if (CaretIsInForStatementCondition(caretPosition, forStatementSyntax))
+                {
+                    end = forStatementSyntax.Condition.FullSpan.End;
+                }
+                else if (CaretIsInForStatementDeclaration(caretPosition, forStatementSyntax))
+                {
+                    end = forStatementSyntax.Declaration.FullSpan.End;
+                }
+                else if (CaretIsInForStatementInitializers(caretPosition, forStatementSyntax))
+                {
+                    end = forStatementSyntax.Initializers.FullSpan.End;
+                }
+                else
+                {
+                    // Should not be reachable because we returned earlier for this case
+                    end = 0;
+                    throw ExceptionUtilities.Unreachable;
+                }
+            }
+            else
+            {
+                end = currentNode.Span.End;
+            }
+
+            // If a node's semicolon is missing, the trailing trivia (including new line) is associated with the last token.
+            // To avoid placing the semicolon on the next line or after comments, we need to adjust the position.
+            return root.FindToken(end).GetPreviousToken().Span.End;
+        }
+
+        private static bool CaretIsInForStatementCondition(int caretPosition, ForStatementSyntax forStatementSyntax)
```

Comment:
>CaretIsInForStatementCondition [](start = 28, length = 30)

Nit: can we use consistent code style for all three `CaretXXX` methods defined here:
1. Use of expression body vs block body.
2. Use of conditional expressions vs if-else statement.
3. If we use if-else statement, please use braces for if-else body.

Reply:
Yes, cleaned this up.  Didn't notice it got so wonky.